nat-lab/tests/test_batching.py

```diff
@@ -0,0 +1,278 @@
+    Batching,
+    LinkDetection,
+    PersistentKeepalive,
+    TelioFeatures,
+    Direct,
+    Wireguard,
+)
+from typing import List
+from utils.batching import PacketFilter, capture_histograms, draw_histogram
+from utils.connection import DockerConnection
+from utils.connection_util import DOCKER_GW_MAP, ConnectionTag, container_id
+
+
+def longest_zero_streak(nums):
+    max_streak = 0
+    current_streak = 0
+
+    for num in nums:
+        if num == 0:
+            current_streak += 1
+            max_streak = max(max_streak, current_streak)
+        else:
+            current_streak = 0
+
+    return max_streak
+
+
+def _generate_setup_parameters(
+    conn_tag: ConnectionTag,
+    adapter: telio.AdapterType,
+    endpoint_providers: List[str],
+    direct_enabled: bool,
+    batching_enabled: bool,
+) -> SetupParameters:
+    features = TelioFeatures(
+        direct=Direct(providers=endpoint_providers) if direct_enabled else None,
+        link_detection=LinkDetection(
+            use_for_downgrade=True, rtt_seconds=1, no_of_pings=1
+        ),
+        wireguard=Wireguard(
+            persistent_keepalive=PersistentKeepalive(
+                direct=70,
+                proxying=70,
+                stun=70,
+                vpn=70,
+            )
+        ),
+        batching=Batching(direct_connection_threshold=35) if batching_enabled else None,
+    )
+
+    item = SetupParameters(
+        connection_tag=conn_tag,
+        adapter_type=adapter,
+        features=features,
+    )
+
+    return item
+
+
+@pytest.mark.batching
+@pytest.mark.asyncio
+@pytest.mark.timeout(300)
+@pytest.mark.parametrize(
+    "setup_params",
+    [
+        pytest.param(
+            [
+                _generate_setup_parameters(
+                    ConnectionTag.DOCKER_CONE_CLIENT_1,
+                    AdapterType.LinuxNativeWg,
+                    ["stun", "upnp", "local"],
+                    direct_enabled=True,
+                    batching_enabled=False,
+                ),
+                _generate_setup_parameters(
+                    ConnectionTag.DOCKER_CONE_CLIENT_2,
+                    AdapterType.LinuxNativeWg,
+                    ["stun", "upnp", "local"],
+                    direct_enabled=True,
+                    batching_enabled=False,
+                ),
+                _generate_setup_parameters(
+                    ConnectionTag.DOCKER_UPNP_CLIENT_1,
+                    AdapterType.LinuxNativeWg,
+                    ["stun", "upnp", "local"],
+                    direct_enabled=True,
+                    batching_enabled=False,
+                ),
+                _generate_setup_parameters(
+                    ConnectionTag.DOCKER_UPNP_CLIENT_2,
+                    AdapterType.LinuxNativeWg,
+                    ["stun", "upnp", "local"],
+                    direct_enabled=True,
+                    batching_enabled=False,
+                ),
+            ],
+            id="no_batching",
```

Comment:
Is it used anywhere?

Reply:
Without the `id` field, the tests look as such:
```
      <Coroutine test_batching_star[CONE_CLIENT_1-LinuxNativeWg-stun-upnp-local-CONE_CLIENT_2-LinuxNativeWg-stun-upnp-local-UPNP_CLIENT_1-LinuxNativeWg-stun-upnp-local-UPNP_CLIENT_2-LinuxNativeWg-stun-upnp-local]>
      <Coroutine test_batching_star[CONE_CLIENT_1-LinuxNativeWg-stun-upnp-local-batch-35-CONE_CLIENT_2-LinuxNativeWg-stun-upnp-local-batch-35-UPNP_CLIENT_1-LinuxNativeWg-stun-upnp-local-batch-35-UPNP_CLIENT_2-LinuxNativeWg-stun-upnp-local-batch-35]>
```

Batching tests should come basically down to one big test that batches, another one which doesn't so I felt it's nice to separate them more clearly

Reply:
These are also incomplete names, as it's only a very small network yet

Reply:
Additionally, I've encountered windows failing with pytest due to too long filenames when saving the logs already in the past